File: src/lib/agents/sub-agents/treks.ts
import { openTripMap } from "@/lib/utils/api-clients";
import { parseOpenTripMapPlace, type Attraction } from "@/lib/utils/parsers";
import { getCached, setCached, attractionCacheKey } from "@/lib/cache/redis";

export interface TreksParams {
  city: string;
}

export async function fetchTreks(params: TreksParams): Promise<Attraction[]> {
  const key = attractionCacheKey(params.city, "treks");
  const cached = await getCached<Attraction[]>(key);
  if (cached) return cached;

  const geoRes = await fetch(
    `${openTripMap.baseUrl}/geoname?name=${encodeURIComponent(params.city)}&apikey=${openTripMap.key()}`
  );
  if (!geoRes.ok) throw new Error(`OpenTripMap geo error: ${geoRes.status}`);
  const geo = await geoRes.json();

  const url = new URL(`${openTripMap.baseUrl}/radius`);
  url.searchParams.set("radius", "30000");
  url.searchParams.set("lon", String(geo.lon));
  url.searchParams.set("lat", String(geo.lat));
  url.searchParams.set("kinds", "natural,hiking,beaches,parks");
  url.searchParams.set("rate", "2");
  url.searchParams.set("limit", "15");
  url.searchParams.set("format", "json");
  url.searchParams.set("apikey", openTripMap.key());

  const res = await fetch(url.toString());
  if (!res.ok) throw new Error(`OpenTripMap treks error: ${res.status}`);
  const places = await res.json();

  const attractions: Attraction[] = places
    .slice(0, 12)
    .map((p: Record<string, unknown>) => ({ ...parseOpenTripMapPlace(p), category: "treks" }));

  await setCached(key, attractions);
  return attractions;
}

File: backend/agents/sub_agents/treks.py
import httpx
from utils.api_clients import OPENTRIPMAP_BASE, opentripmap_key
from utils.parsers import Attraction, parse_opentripmap_place
from cache.redis import get_cached, set_cached, attraction_cache_key


async def fetch_treks(city: str) -> list[Attraction]:
    key = attraction_cache_key(city, "treks")
    cached = await get_cached(key)
    if cached:
        return [Attraction(**a) for a in cached]

    async with httpx.AsyncClient(timeout=15) as client:
        geo_res = await client.get(
            f"{OPENTRIPMAP_BASE}/geoname",
            params={"name": city, "apikey": opentripmap_key()},
        )
        geo_res.raise_for_status()
        geo = geo_res.json()

        res = await client.get(
            f"{OPENTRIPMAP_BASE}/radius",
            params={
                "radius": "30000",
                "lon": geo["lon"],
                "lat": geo["lat"],
                "kinds": "natural,hiking,beaches,parks",
                "rate": "2",
                "limit": "15",
                "format": "json",
                "apikey": opentripmap_key(),
            },
        )
        res.raise_for_status()
        places = res.json()

    attractions = [
        Attraction(**{**parse_opentripmap_place(p).__dict__, "category": "treks"})
        for p in places[:12]
    ]

    await set_cached(key, [a.to_dict() for a in attractions])
    return attractions
